### ISSUE TITLE
Plinth design principles?

### ISSUE THREAD

**Issue by JoKeyser** [freedombox/Plinth#992]
# Plinth design principles?
From my experience as new contributor/reviewer, there's many subtle design decisions to make and evaluate. I think that "Plinth design principles" would be useful, that bridge from the project's mission to the implementation details. I don't mean to repeat existing work, e.g. (partially outdated?) design ideas documented in the wiki](https://wiki.debian.org/FreedomBox/Design) (some pre-date Plinth), and the *graphical* user design plans like the [card-based UI](https://github.com/freedombox/Plinth/issues/739). Instead, the goal of this thread would be to arrive at a consensus of actionable design principles that guide our implementation. I know some already exist (mainly in our heads, maybe elsewhere?), we would "just" have to articulate them explicitly, and condense them into one place.

Below, I first clarify further what I mean, followed by proposed candidate principles.

## From mission, to strategy, to (imo missing) Plinth design principles, to implementation
To preemptively clarify this (potential) discussion, let me define some "layers" of the project's setup. Please let me know if something is wrong, or misleading, or missing.

### Our mission (only for clarification)
Our ***mission*** is articulated in multiple places, e.g. [the wiki](https://wiki.debian.org/FreedomBox). In brief, we want to simplify self-hosting of privacy-respecting Internet services, to avoid the proprietary cloud. Note that this goal is shared with many great projects, such as [YunoHost](https://yunohost.org/), or [[mail-in-a-box](https://mailinabox.email/), or [Souvereign](https://github.com/sovereign/sovereign), or [Sandstorm(?)](https://sandstorm.io/).

### Our strategy (mainly for clarification)
We start differing from other projects with our ***strategy***; here my summary. (Note there have been some varying proposed strategies in the past, see [design wiki page](https://wiki.debian.org/FreedomBox/Design).)

* We're a Debian pure blend, targeting people without *special* technical skills.
    * About skills, I think avoidable ones include Debian tools/logic, SSH, and editing config files.
      However, *un*avoidable skills include (very) basic knowledge about IP-routing, domains, and cryptographic certificates.
      Documentation of the required basics are part of the project; setups beyond this scope are not, so we may point people to other projects and resources.
    * We're a 2nd filter from all of FLOSS => Debian -> FreedomBox... to the end user.
      We curate Debian's output with simple UIs (Plinth, Android app), by selecting packages, and by setting sensible defaults, thus exposing only (mission-)critical configuration.
    * To reach an **x**-fold larger audience (than plain Debian), we need to create an **x**-fold simpler interface.
      On the flip side, we take away some choices, and power users might be better off with other projects that better accomodate heavy customization.
    * We augment functionality of Debian packages by integrating them (e.g. Single-Sign-On, or using LE certs with many services).
    * As part of Debian, we interact directly with its infrastructure and send bugs or patches to the package maintainers instead of fixing things "only for us".
* Translating UI and documentation is essential for widespread use.
* Stability of the system is more important than the number of features, I guess.
* To prioritize our efforts (e.g. avoid feature creep), we articulate [user stories](https://wiki.debian.org/FreedomBox/UserExperience) (are there better links?).

### Plinth design principles? (What this thread should be about.)
If we agree, our ***Plinth design principles*** would take the strategy for granted and break it down, such that our implementations are effectively following it. They would serve as reference for authors and reviewers, and thus help in creating a unified interface, and maybe make the code more maintainable. See below for some proposed candidates.

### Implementation (only for clarification)
By implementation I mean the software, docs, graphics, and websites we actually create. Plinth is written in the Python Django framework, to trigger so-called *actions* that change the system state. Also these implementation decisions are *not* part of what I want to discuss, but rather *how* to use them.

## Proposed Plinth design principles
Here a list of candidate principles that we might discuss and extend. Letters denote a principle, and numbers are (loosely) resulting, concrete implementation decisions.

* A. All of Plinth's action scripts have to abide by the contract defined in https://github.com/freedombox/Plinth/blob/master/plinth/actions.py#L20.
* B. Maximize transparency for administration via a command line interface (cli).
   Put reversely, minimize interference with knowledgeable admins and advanced tools [e.g. cockpit](http://cockpit-project.org/).
    * 1) Plinth's actions should have a consistent, documented cli.
    * 2) Plinth's actions and buttons in the UI should have close correspondence.
    * 3) Avoid to store settings or system states in ways that are unique to Plinth.
        Instead, try to read/write configurations/states from/into files that are easily cli-accessible.
    * 4) If a Plinth action changes a setting, it should become effective immediately.
        E.g. if a service is already running, send a signal to re-load its config (or restart it).
* C. Keep configuration of packages upgradable by `unattended-upgrades`.
   This is currently being violated, probably worth its own issue; Plinth often directly edits config files, preventing future unattended upgrades. The problem has been acknowledged before [here in the design wiki](https://wiki.debian.org/FreedomBox/Design/BoxConfiguration). Maybe there exists a universal solution? Maybe we need to use several approaches, depending on the package. For example, @jvalleroy and Jonas are [working on a solution for radicale](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=848064); maybe this can be applied to others? Maybe [propellor (with apt property)](https://hackage.haskell.org/package/propellor-4.7.7/docs/Propellor-Property-Apt.html) [(available in sid)](https://packages.debian.org/sid/propellor) would be a general solution..?
* D. Every action is covered by an automatic test. A proposed patch only gets integrated after it passes all tests. Again, this is not quite achieved yet. But since stability is very important, we should be checking everything as early as possible in the development process.
* E. Features are designed first, before being implemented.
   Maybe harder to achieve at all times, but worthwile to avoid duplicated efforts to correct bad design/implementation decisions. Could be practically as simple as always creating an issue first, to give an opportunity to comment on, and define what the best solution is, and implement only if it got the `ready-for-implementation` tag.
* F. ...?

**Comment by JoKeyser** [comment]
Another proposal:
Avoid referring to "Plinth" in interfaces; instead use "FreedomBox" (or the abstraction box_name).
Pllinth is an unnecessary additional term for users that might cause confusion.

**Comment by JosephKiranBabu** [comment]
@JoKeyser Maybe we should rename the repositories as well, Plinth to freedombox-web and the Android app to freedombox-android

**Comment by JoKeyser** [comment]
@JosephKiranBabu I agree that this would reduce confusion. I actually have no idea why it's called Plinth, historically :).

**Comment by JoKeyser** [comment]
Another quote from @SunilMohanAdapa : "When we use a utility method from Plinth, we can trust it to give expected values. Otherwise, we should fix the method to be more robust."
That's a good idea, I often get too paranoid when adding stuff, and add unnecesary checks when calling a function I'm not too familiar with.
